### ISSUE TITLE
Lungs Segmentation | long term

### ISSUE THREAD

**Issue by vessemer** [concept-to-clinic/concept-to-clinic#138]
The initial lungs segmentation method has been provided by the PR [#133](https://github.com/concept-to-clinic/concept-to-clinic/pull/133) due to the issue [#120](https://github.com/concept-to-clinic/concept-to-clinic/issues/120). The provided approach, however, can be further improved e.g., by reducing its trend to false positives and made the algorithm more stable. Furthermore, the ability to correctly separate the tissues such as bones, lungs, bronchial, etc. will be beneficial for a further work with data augmentation.

## Possible Solutions
For example, the work of [van Rikxoort et al.](https://www.ncbi.nlm.nih.gov/pubmed/19673192) describes the automatic error detection method via the convex hull complement to a coastline of lungs:  

![](https://preview.ibb.co/nB7ib5/image.png)  

Furthermore, the method provided by [S. Hu et al.](http://ieeexplore.ieee.org/abstract/document/929615/) is aimed at junction line enhancement followed by lungs separation which I've found to be unreasonable resource consumptive though.

![](https://preview.ibb.co/dxVRDk/Screenshot_from_2017_09_26_15_10_26.png)

 The ability of the bronchial / lungs separation described in the paper of [T. Kitasaka et al.](http://campar.in.tum.de/pub/kitasaka2010pulmo/kitasaka2010pulmo.pdf) I guess will also be valuable as an additional instrument of data augmentation.

![](https://image.ibb.co/m2x9zQ/Screenshot_from_2017_09_26_15_17_37.png)

> Volume rendered results of the ground truth (left), the proposed method (middle) and the [adaptive branch tracing method](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.164.7783&rep=rep1&type=pdf) (right). Yellow region indicates TP while blue region shows FP Though the bronchial lumen is interrupted by a tumor indicated by the arrow, the proposed method can extract the lumen regions beyond the tumor.   
  
P.S. Of course, contributions like a radiologist's handcrafted annotations of anatomical structures inside CT volume will be appreciated.

## Current Behavior
The lungs segmentation lied in the `preprocess/lung_segmentation.py` consists of the followed steps:
Step 1: Convert into a binary image.
Step 2: Remove the blobs connected to the border of the image.
Step 3: Label the image.
Step 4: Keep the labels with 2 largest areas.
Step 5: Erosion operation with a disk of radius 2. To separate lung nodules attached to blood vessels.
Step 6: Closure operation with a disk of radius 10. To keep nodules attached to a lung wall.
Step 7: Fill in the small holes inside the binary mask of lungs.

## Acceptance criteria
The good measure of a lungs segmentation algorithm should be [Hausdorff distance](https://en.wikipedia.org/wiki/Hausdorff_distance) which has efficient implementation by [`scipy.spatial.distance.directed_hausdorff`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.directed_hausdorff.html#scipy.spatial.distance.directed_hausdorff), thus the acceptance criteria will rely on the Hausdorff tests. Nonetheless, this is the long-term issue, thus different type of contribution may be accepted by this issue.

**Comment by WGierke** [comment]
Yep, we're definitely supposed to refine the lung segmentation. Have a look at e.g. patient 0001 and the nodule at 315.04852321,  365.87447257, -116.12078059 (slice ~91):
![screenshot from 2017-09-30 10-12-50](https://user-images.githubusercontent.com/6676439/31044067-217e7ba6-a5c8-11e7-8644-7db092746b47.png)
It's state is labeled as solid, which isn't included in the current lung segmentation:
![screenshot from 2017-09-30 10-12-22](https://user-images.githubusercontent.com/6676439/31044079-4dd01ec6-a5c8-11e7-9b5b-714c1e7c0ec8.png)
However, some of the slices of the nodule are included in the lung segmentation but currently this would still falsify statistics such as the predicted volume of a nodule as to be implemented in #3 .

**Comment by kmader** [comment]
As part of the evaluation, the [Luna16](https://luna16.grand-challenge.org/data/) data include right and left segmented lungs with trachea which could serve as a good baseline

**Comment by caseyfitz** [comment]
New issue #283 focused on nodule segmentation and volume estimation (as opposed to segmentation of the entire lung).
